### PR TITLE
Python 3.12 support, compatability with Eliot 1.15

### PR DESCRIPTION
**Pull request by itamarst**
Fixes https://tahoe-lafs.org/trac/tahoe-lafs/ticket/4071 (not sure where 3072 in branch name came from)

* [x] Tests pass
* [x] Tests run on 3.12 on Windows. I chose to use GitHub because CircleCI don't bother documenting what versions of Python they have on their image, and since it's so new I don't have a lot of confidence they've added it, so don't want to waste time on it this week.

### CODE REVIEW

File: src/allmydata/immutable/upload.py [hunk 1]
```diff
@@ -91,7 +91,7 @@ def _serialize_existing_shares(existing_shares):
 
 def _serialize_happiness_mappings(happiness_mappings):
     return {
-        sharenum: base32.b2a(serverid)
+        str(sharenum): ensure_str(base32.b2a(serverid))
```
Comment (meejah) [inline src/allmydata/immutable/upload.py:94]: `str` -> `ensure_str` ...? (or, why not?

File: src/allmydata/test/test_storage_http.py [hunk 2]
```diff
@@ -442,6 +442,9 @@ def test_client_side_schema_validation(self):
             result_of(client.get_version())
 
     @given(length=st.integers(min_value=1, max_value=1_000_000))
+    # On Python 3.12 we're getting weird deadline issues in CI, so disabling
+    # for now.
```
Comment (meejah) [inline src/allmydata/test/test_storage_http.py:446]: hmm, doesn't the "ci" Hypothesis profile already turn off the deadline? (or is "HealthCheck.too_slow" different from this deadline?)

**Review by meejah — APPROVED** [review]
LGTM, like less code!
(Some inline comments, optional)

File: src/allmydata/immutable/upload.py (hunk 1 above)
Comment (itamarst) [reply src/allmydata/immutable/upload.py:94]:
```
>>> b = b"abc"
>>> str(b)
"b'abc'"
```

File: src/allmydata/test/test_storage_http.py (hunk 2 above)
Comment (itamarst) [reply src/allmydata/test/test_storage_http.py:446]: I don't know :(

File: src/allmydata/immutable/upload.py (hunk 1 above)
Comment (meejah) [reply src/allmydata/immutable/upload.py:94]: sorry, I guess I wasn't clear: why `str(sharenum)` and not `ensure_str(sharenum)`...?

Comment (itamarst) [reply src/allmydata/immutable/upload.py:94]: sharenum is an int so it's just a matter of making sure keys are strings, vs dealing with bytes.

**Pull request merged**